cancat/uds.py

```diff
@@ -8,39 +8,61 @@ def __init__(self, c, tx_arbid, rx_arbid):
         self.c = c
         self.tx_arbid = tx_arbid
         self.rx_arbid = rx_arbid
-        self.t = threading.Thread(target = self.SendTesterPresent)
-        self.TesterPresent = False
-        self.t.setDaemon(True)
-        self.t.start()
 
-    def ReassembleIsoTP(self, start_index = 0):
+    def ReassembleIsoTP(self, start_index = 0, service = None):
         msg_found = False
         while msg_found is False:
             for idx, ts, arbid, msg in self.c.genCanMsgs(start=start_index, arbids=[self.rx_arbid]):
-                msg_found = True
-                return hex(arbid), msg.encode('hex')
+                # Check that the message is for the expected service, if specified
+                if service is not None:
+                    # Check if this is the right service, or there was an error
+                    if ord(msg[1]) == service or ord(msg[1]) == 0x7f:
+                        msg_found = True
+                        return arbid, msg
+                else:
+                    msg_found = True
+                    return arbid, msg
             time.sleep(0.1)
 
     def SendTesterPresent(self):
-        while True:
-            if self.TesterPresent:
-                self.c.CANxmit(self.tx_arbid, "023E000000000000".decode('hex'))
+        while self.TesterPresent is True:
+            self.c.CANxmit(self.tx_arbid, "023E000000000000".decode('hex'))
             time.sleep(2.0)
 
     def StartTesterPresent(self):
         self.TesterPresent = True
+        self.t = threading.Thread(target = self.SendTesterPresent)
```

Comment:
There is no error message. I'm trying to solve two different problems:

1. During development, I frequently reload the module after making a change. I needed a way to stop the thread running before I lost my reference to the old UDS object. Otherwise I wind up in a situation where I have 50 zombie threads sending the TesterPresent signal.

2. Before I set the thread as a Daemon, when you exited CanCat (Ctrl-d) python wouldn't actually stop since that thread was still running. The atexit function may be a better solution than the Daemon.